src/contracts/mocks/nf-token-metadata-enumerable-mock.sol

```diff
@@ -60,4 +60,120 @@ contract NFTokenMetadataEnumerableMock is
     super._burn(_tokenId);
   }
 
+  /**
+   * @dev Mints a new NFT.
+   * @notice This is an internal function which should be called from user-implemented external
+   * mint function. Its purpose is to show and properly initialize data structures when using this
+   * implementation.
+   * @param _to The address that will own the minted NFT.
+   * @param _tokenId of the NFT to be minted by the msg.sender.
+   */
+  function _mint(
+    address _to,
+    uint256 _tokenId
+  )
+    internal
+    override(NFToken, NFTokenEnumerable)
+    virtual
+  {
+    super._mint(_to, _tokenId);
+    tokens.push(_tokenId);
+    idToIndex[_tokenId] = tokens.length - 1;
+  }
+
+  /**
+   * @dev Burns a NFT.
+   * @notice This is an internal function which should be called from user-implemented external
+   * burn function. Its purpose is to show and properly initialize data structures when using this
+   * implementation. Also, note that this burn implementation allows the minter to re-mint a burned
+   * NFT.
+   * @param _tokenId ID of the NFT to be burned.
+   */
+  function _burn(
+    uint256 _tokenId
+  )
+    internal
+    override(NFTokenMetadata, NFTokenEnumerable)
+    virtual
+  {
+    super._burn(_tokenId);
+
+    uint256 tokenIndex = idToIndex[_tokenId];
+    uint256 lastTokenIndex = tokens.length - 1;
+    uint256 lastToken = tokens[lastTokenIndex];
+
+    tokens[tokenIndex] = lastToken;
+
+    tokens.pop();
+    // This wastes gas if you are burning the last token but saves a little gas if you are not.
+    idToIndex[lastToken] = tokenIndex;
+    idToIndex[_tokenId] = 0;
+  }
```

Comment:
What about calling: 

    NFTokenEnumerable._burn(_tokenId);
    NFTokenMetadata._burn(_tokenId);

That works in practice but I'm not sure why since in theory super._burn should be called twice and fail because of that. 

Another option would be:
NFTokenEnumerable._burn(_tokenId);
+ the missing code you pasted.

Reply:
@fulldecent Can you check this out?

Reply:
This implementation

```solidity
NFTokenEnumerable._burn(_tokenId);
NFTokenMetadata._burn(_tokenId);
```

would, as you noted, call `NFToken._burn` twice. This emits two `Transfer` events which is not allowed by specification.

---

Yes, I agree that

```solidity
NFTokenEnumerable._burn(_tokenId);
if (bytes(idToUri[_tokenId]).length != 0)
{
    delete idToUri[_tokenId];
}
```

is a good approach.